packages/flutter/test/widgets/editable_text_test.dart

```diff
@@ -2080,6 +2081,150 @@ void main() {
+
+    final Widget widget = MaterialApp(
+      home: Align(
+        alignment: Alignment.topLeft,
+        child: Container(
+          child: SizedBox(
+            width: 100,
+            child: EditableText(
+              key: editableTextKey,
+              controller: controller,
+              focusNode: FocusNode(),
+              style: Typography(platform: TargetPlatform.iOS).black.subhead,
+              cursorColor: Colors.blue,
+              backgroundCursorColor: Colors.grey,
+              selectionControls: cupertinoTextSelectionControls,
+              keyboardType: TextInputType.text,
+            ),
+          ),
+        ),
+      ),
+    );
+
+    await tester.pumpWidget(widget);
+
+    final EditableTextState state =
+        tester.state<EditableTextState>(find.byType(EditableText));
+    final RenderEditable renderEditable = state.renderEditable;
+    final Scrollable scrollable = tester.widget<Scrollable>(find.byType(Scrollable));
+
+    bool leftVisibleBefore = false;
+    bool rightVisibleBefore = false;
+
+    Future<void> verifyVisibility(
+      bool leftVisible,
+      Symbol leftPosition,
+      bool rightVisible,
+      Symbol rightPosition,
+    ) async {
+      await tester.pump();
+
+      // Check the signal from RenderEditable about whether they're within the
+      // viewport.
+
+      expect(renderEditable.selectionStartInViewport.value, equals(leftVisible));
+      expect(renderEditable.selectionEndInViewport.value, equals(rightVisible));
+
+      // Check that the animations are functional and going in the right
+      // direction.
+
+      final List<Widget> transitions =
+        find.byType(FadeTransition).evaluate().map((Element e) => e.widget).toList();
+      final FadeTransition left = transitions[0];
+      final FadeTransition right = transitions[1];
+
+      if (leftVisibleBefore)
+        expect(left.opacity.value, equals(1.0));
+      if (rightVisibleBefore)
+        expect(right.opacity.value, equals(1.0));
+
+      await tester.pump(TextSelectionOverlay.fadeDuration ~/ 2);
+
+      if (leftVisible != leftVisibleBefore)
+        expect(left.opacity.value, equals(0.5));
+      if (rightVisible != rightVisibleBefore)
+        expect(right.opacity.value, equals(0.5));
+
+      await tester.pump(TextSelectionOverlay.fadeDuration ~/ 2);
+
+      if (leftVisible)
+        expect(left.opacity.value, equals(1.0));
+      if (rightVisible)
+        expect(right.opacity.value, equals(1.0));
+
+      leftVisibleBefore = leftVisible;
+      rightVisibleBefore = rightVisible;
+
+      // Check that the handles' positions are correct (clamped within the
+      // viewport but not stuck).
+
+      final List<Positioned> positioned =
+        find.byType(Positioned).evaluate().map((Element e) => e.widget).cast<Positioned>().toList();
+
+      final Size viewport = renderEditable.size;
+
+      void testPosition(double pos, Symbol expected) {
+        if (expected == #left)
```

Comment:
These symbols might be a little easier to understand if they were called "leftViewportEdge" or "withinViewport" etc